[mlir][LLVM] Add support for constant struct with multiple fields

Lancern opened the pull request:
Currently `mlir.llvm.constant` of structure types restricts that the structure type effectively represents a complex type -- it must have exactly two fields of the same type and the field type must be either an integer type or a float type.

This PR relaxes this restriction and it allows the structure type to have an arbitrary number of fields.

Lancern committed:
[mlir][LLVM] Add support for constant struct with multiple fields

Currently `mlir.llvm.constant` of structure types restricts that the structure
type effectively represents a complex type -- it must have exactly two fields
of the same type and the field type must be either an integer type or a float
type.

This patch relaxes this restriction and it allows the structure type to have an
arbitrary number of fields.

Dinistro commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2678,31 +2678,30 @@ LogicalResult LLVM::ConstantOp::verify() {
-      return emitOpError()
-             << "expected array attribute with two elements of the same type";
+
+    ArrayRef<Type> elementTypes = structType.getBody();
```
Shouldn't this check now be recursive, given that one can define nested struct constants?

Dinistro commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2678,31 +2678,30 @@ LogicalResult LLVM::ConstantOp::verify() {
-                "integer type";
+    for (size_t i = 0; i < elementTypes.size(); ++i) {
+      auto element = arrayAttr[i];
+      if (!mlir::isa<IntegerAttr, FloatAttr>(element)) {
```
```suggestion
      if (!isa<IntegerAttr, FloatAttr>(element)) {
```
This namespace is used. Same for other occurrences.

Lancern replied:
Updated.

Dinistro commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -557,20 +557,20 @@ llvm::Constant *mlir::LLVM::detail::getLLVMConstant(
-    if (!imag)
-      return nullptr;
-    return llvm::ConstantStruct::get(structType, {real, imag});
+    llvm::SmallVector<llvm::Constant *, 8> structElements;
```
```suggestion
   SmallVector<llvm::Constant *, 8> structElements;
```

Lancern replied:
Updated.

Dinistro commented on mlir/test/Dialect/LLVMIR/invalid.mlir:
Please add a test with a nested struct

Lancern replied:
As suggested by @gysit :

> we could initially support structs without pointers and without nesting.

Should we include support for nested structure constant in this PR?

Dinistro reviewed (commented):
Thanks for the change. I dropped a few comments regarding nested structs.

gysit reviewed (commented):
Can you elaborate on the use case for this change? 

I think the constant operation is only meant to produce constants of non-aggregate types. The complex type is an exception bcs there is some support for builtin complex attributes which produces an array attribute with two elements.

Note that proper struct support is not possible bcs a struct can contain pointers which cannot be modeled as attributes. Additionally, they can be nested, which may be modeled with attributes but things get complex.

Instead if you want to produce a constant struct, the preferred way is to use the undef operation and then set the individual fields of the struct. 

You can test this by writing some example:
```
define { i32, i32 } @test() {
   ret {i32, i32} { i32 -1, i32 0}
}
```
And then call  `./bin/mlir-translate --import-llvm ./test.ll` to import into MLIR:
```
  llvm.func @test() -> !llvm.struct<(i32, i32)> {
    %0 = llvm.mlir.constant(0 : i32) : i32
    %1 = llvm.mlir.constant(-1 : i32) : i32
    %2 = llvm.mlir.undef : !llvm.struct<(i32, i32)>
    %3 = llvm.insertvalue %1, %2[0] : !llvm.struct<(i32, i32)> 
    %4 = llvm.insertvalue %0, %3[1] : !llvm.struct<(i32, i32)> 
    llvm.return %4 : !llvm.struct<(i32, i32)>
  }
```
This represents an struct constant in LLVM dialect. While this particular example could be supported things will fail as soon as there are pointers:
```
  llvm.func @test() -> !llvm.struct<(ptr, i32)> {
    %0 = llvm.mlir.constant(0 : i32) : i32
    %1 = llvm.mlir.zero : !llvm.ptr
    %2 = llvm.mlir.undef : !llvm.struct<(ptr, i32)>
    %3 = llvm.insertvalue %1, %2[0] : !llvm.struct<(ptr, i32)> 
    %4 = llvm.insertvalue %0, %3[1] : !llvm.struct<(ptr, i32)> 
    llvm.return %4 : !llvm.struct<(ptr, i32)>
  }
```
I would thus stick to the current limitations of constant op and represent aggregate constants using a sequence of operations. Otherwise there will be multiple ways of representing the same thing. That being said maybe you have a good argument why you need constant op to support arrays?

Lancern commented:
@gysit Thanks for your helpful comment on this topic!

> Can you elaborate on the use case for this change?

Well my actual use case is rather minor compared to the proposed changes here. I'm working on [clangir](https://github.com/llvm/clangir) which provides an MLIR dialect for modeling C/C++ language semantics, and recently I'm trying to work out the appropriate LLVMIR for [pointers to member functions (P2MF)](https://en.cppreference.com/w/cpp/language/pointer#Pointers_to_member_functions). The layout of a P2MF is ABI-specific and currently we're focusing on the Itanium ABI which defines the layout as something like `struct { intptr_t, ptrdiff_t }`. The two members are both integers, but their types may differ. I tried directly lower a P2MF constant into `llvm.mlir.constant` but I come across the current limitation on structure constants. That's the full story.

> I think the constant operation is only meant to produce constants of non-aggregate types.

This makes a lot sense and the pointer problem indeed renders a complete support for arbitrary struct impossible. But for the pointer problem, IMHO it is not really related to the changes here -- you can't produce a constant pointer scalar via `llvm.mlir.constant` either, could you? LLVM itself does not have pointer constants (other than null pointer constants) and that's a more related problem IMO. Besides, `llvm.mlir.constant` is already being used for creating aggregates like vectors (but not arrays or structs yet). So one of the motivation of this PR is that I would expect `llvm.mlir.constant` to be able to produce any constant value supported by LLVM.

gysit commented:
> The layout of a P2MF is ABI-specific and currently we're focusing on the Itanium ABI which defines the layout as something like struct { intptr_t, ptrdiff_t }.  The two members are both integers, but their types may differ.

I would expect that once you are at the LLVM dialect level you probably want to know the size of these integers (e.g. by querying the data layout)? So from an LLVM dialect point of view the problem is about a simple struct with two integer elements?

> This makes a lot sense and the pointer problem indeed renders a complete support for arbitrary struct impossible. But for the pointer problem, IMHO it is not really related to the changes here -- you can't produce a constant pointer scalar via llvm.mlir.constant either, could you? LLVM itself does not have pointer constants (other than null pointer constants) and that's a more related problem IMO. Besides, llvm.mlir.constant is already being used for creating aggregates like vectors (but not arrays or structs yet). So one of the motivation of this PR is that I would expect llvm.mlir.constant to be able to produce any constant value supported by LLVM.

A nullptr is represented by:
`%1 = llvm.mlir.zero : !llvm.ptr`
There are similar operations for undef / poison values (not just pointers).

I am not completely opposed to supporting constant structs. If we do it, we need have a proper verifier and the documentation should clearly specify what kind of structs are supported. E.g. we could initially support structs without pointers and without nesting. 

I just wanted to point out that there are already alternative ways to express struct constants that work out of the box and that support arbitrary structs. The advantage of your PR is mainly that it simplifies generating struct constants and that folding is potentially easier.

gysit commented:
FYI, there is a related PR in flight https://github.com/llvm/llvm-project/pull/102906.

Lancern commented:
> So from an LLVM dialect point of view the problem is about a simple struct with two integer elements?

Yes, that's my use case.

> If we do it, we need have a proper verifier and the documentation should clearly specify what kind of structs are supported. E.g. we could initially support structs without pointers and without nesting.

I agree with this. I updated the documentation and listed the constraints on the LLVM struct type of the result constant.

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -557,20 +557,20 @@ llvm::Constant *mlir::LLVM::detail::getLLVMConstant(
-    if (!imag)
-      return nullptr;
-    return llvm::ConstantStruct::get(structType, {real, imag});
+    SmallVector<llvm::Constant *, 8> structElements;
```
```suggestion
    SmallVector<llvm::Constant *> structElements;
```
nit: We usually do not provide a size except there is a specific reason.

Lancern replied:
Updated.

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -557,20 +557,20 @@ llvm::Constant *mlir::LLVM::detail::getLLVMConstant(
-    return llvm::ConstantStruct::get(structType, {real, imag});
+    SmallVector<llvm::Constant *, 8> structElements;
+    structElements.reserve(structType->getNumElements());
+    for (size_t i = 0; i < arrayAttr.size(); ++i) {
```
```suggestion
    for (auto [elemType, elemAttr] : zip_equal(arrayAttr, structType.getElementTypes()))
```
nit: Would it be possible to use zip_equal here?

Lancern replied:
Updated.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2710,32 +2710,31 @@ LogicalResult LLVM::ConstantOp::verify() {
+    for (size_t i = 0; i < elementTypes.size(); ++i) {
+      auto element = arrayAttr[i];
+      if (!isa<IntegerAttr, FloatAttr>(element)) {
+        return emitOpError() << "expected struct element types to be floating "
```
Can you add a test for this?

Lancern replied:
Test added.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2710,32 +2710,31 @@ LogicalResult LLVM::ConstantOp::verify() {
-             << "expected struct element types to be floating point type or "
-                "integer type";
+    for (size_t i = 0; i < elementTypes.size(); ++i) {
+      auto element = arrayAttr[i];
```
```suggestion
      Attribute element = arrayAttr[i];
```
nit: here it makes sense to spell the type since it is not on the rhs.

Lancern replied:
Updated.

gysit reviewed (commented):
Thanks for the updates! I have added some comments.

One thing we do not yet check are llvm.struct types nested an llvm.array. I wonder if such a verification could be added without breaking exiting users/tests. If yes, that would be nice to have. However, I believe this is currently not verified either.

Lancern commented:
> One thing we do not yet check are llvm.struct types nested an llvm.array.

Updated the verifier and added a check that ensures every element type within a struct type is either int or float. Also added a test case for this.

gysit commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2710,32 +2711,38 @@ LogicalResult LLVM::ConstantOp::verify() {
+               << "expected struct element attribute types to be floating "
+                  "point type or integer type";
+      }
+      auto elementType = cast<TypedAttr>(element).getType();
```
```suggestion
      auto elementType = dyn_cast<TypedAttr>(element).getType();
```
Didn't see this before. I think we should dyn_cast here (as it was before). In case the attribute does not implement the TypedAttr interface we want to return an error and not assert.

gysit replied:
True you are right!

Then this PR is ready to go. Do you have commit rights?

Lancern replied:
> Do you have commit rights?

No.

gysit replied:
Ok I will land it then.

gysit reviewed (approved):
Nice!

LGTM modulo one last place where I believe a dyn_cast is necessary.
